src/kmer_collection.h

```diff
@@ -136,47 +165,70 @@ class KmerCollection {
     }
   }
 
+  /**
+  * Generates kmers if necessary.
+  * \return Number of kmers.
+  */
+  int size() {
+    generateKmers();
+    return m_kmerListSize;
+  }
+
+  /**
+  * Generates kmers if necessary.
+  * \return Iterator to KmerCollection's first element.
+  */
   KmerCollection::iterator begin() {
+    generateKmers();
+    return KmerIterator(this, 0);
+  }
+
+  /**
+  * \return Iterator to the element after KmerCollection's last element.
+  */
+  inline KmerCollection::iterator end() {
+    return KmerIterator(this, m_kmerListSize);
+  }
+
+ private:
+  inline void generateKmers() {
     if(m_kmerList == nullptr) {
       chooseGenerator();
 
+      m_kmerListSize = (m_seqEntry->seq().size() - m_kmerSize) / m_stepSize + 1;
+
       m_kmerList = new uint_fast64_t[m_kmerListSize];
       const char* ptr = m_seqEntry->seq().data();
 
       for(int i = 0; i < m_kmerListSize; ++i) {
-        m_kmerList[i] = (this->*m_usedGenerator)(&(ptr[i]));
+        m_kmerList[i] = (this->*m_usedGenerator)(&(ptr[i*m_stepSize]));
       }
     }
-    return KmerIterator(this, 0);
   }
 
-  KmerCollection::iterator end() {
-    return KmerIterator(this, m_kmerListSize);
-  }
-
- private:
-  void chooseGenerator() {
-    if(m_compress == false) switch(m_kmerSize) {
-      case 8: m_usedGenerator = &KmerCollection::generateUncompressedKmers8; break;
-      default: m_usedGenerator = &KmerCollection::generateUncompressedKmersN; break;
+  inline void chooseGenerator() {
+    if(!m_compress) switch(m_kmerSize) {
+      case 8: m_usedGenerator = &KmerCollection::generateUncompressedKmer8; break;
+      default: m_usedGenerator = &KmerCollection::generateUncompressedKmerN; break;
     }
     else switch(m_kmerSize) {
-      case 8: m_usedGenerator = &KmerCollection::generateKmers8; break;
-      default: m_usedGenerator = &KmerCollection::generateKmersN; break;
+      case 8: m_usedGenerator = &KmerCollection::generateKmer8; break;
+      default: m_usedGenerator = &KmerCollection::generateKmerN; break;
     }
   }
 
   // This assumes the next 8 chars is readable
-  uint_fast64_t generateKmers8(const char* ptr) {
-    uint_fast64_t kmer = *(const uint_fast64_t*)ptr;
+  uint_fast64_t generateKmer8(const char* ptr) const {
+    const uint_fast64_t mask = 0x0606060606060606;
+    uint_fast64_t kmer = *(const uint_fast64_t*)ptr & mask;
```

Comment:
Don't use c-style casts and don't assume `uint_fast64_t` is guaranteed to be 64 bit.